components/concept/storage/src/main/java/mozilla/components/concept/storage/LoginsStorage.kt

```diff
@@ -220,10 +218,22 @@ interface LoginValidationDelegate {
      *
      * Note that this method is not thread safe.
      *
-     * @returns a [LoginValidationDelegate.Result], detailing whether [login] can be saved as a new
-     * value, used to update an existing one, or an error occured.
+     * @returns a [LoginValidationDelegate.Result], detailing whether [login] can be saved or updated,
+     * or an error occurred.
      */
     fun validateCanPersist(login: Login): Deferred<Result>
+
+    /**
+     * A type of persistence operation, either 'create' or 'update'.
+     */
+    enum class Operation { CREATE, UPDATE }
+
+    /**
+     * Checks whether a [login] should be saved or updated.
+     *
+     * @returns a [LoginValidationDelegate.Operation], detailing whether [login] should be saved or updated
+     */
+    fun getPersistenceOperation(newLogin: Login): Deferred<Operation>
```

Comment:
I feel like you explained this to me during our conversation and I'm just being dense, but this seems to cover a really similar conceptual space to `validateCanPersist`.  Why can't they be consolidated into one function?

Reply:
So the way I was thinking about it was from the UI perspective I want to separate the logic of "should we save or update?" (`getPersistenceOperation`) and "CAN we save or update this login?" (`validateCanPersist`). In `LoginDialogFragment` these are used for the button text and the error states/enabled states respectively. Does that make sense?

Reply:
I understand the distinction from the UI perspective, but I'm not sure why they can't rely on the same underlying check.  If it's possible to consolidate them, I think the code would be easier to follow.

Reply:
Maybe - I think this would turn into a 70+ line behemoth function though and I see them as fundamentally different logic. I can look again tomorrow to see what this could look like.

Reply:
Okay attempted! Let me know if you like this better. It's fairly large but maybe more understandable?